proxy/src/control/cache.rs

```diff
@@ -93,41 +137,81 @@ impl<T> Cache<T> where T: Clone + Copy + Eq + std::hash::Hash {
         self.reset_on_next_modification = false;
     }
 
-    pub fn clear<F>(&mut self, on_change: &mut F) where F: FnMut(T, CacheChange) {
-        self.retain(&HashSet::new(), on_change)
+    pub fn clear<F>(&mut self, on_change: &mut F)
+    where
+        F: FnMut((K, V), CacheChange),
+    {
+        self.update_intersection(&IndexMap::new(), on_change)
     }
 
-    pub fn retain<F>(&mut self, to_retain: &HashSet<T>, mut on_change: F)
-        where F: FnMut(T, CacheChange)
+    /// Update the cache to contain the intersection of its current contents
+    /// and the key-value pairs in `to_update`. Pairs not present in
+    /// `to_update` will be removed from the cache, and any keys in the cache
+    /// with different values from those in `to_update` will be ovewritten to
+    /// match `to_update`.
+    pub fn update_intersection<F, Q>(
+        &mut self,
+        to_update: &IndexMap<K, V>,
+        mut on_change: F
+    )
+    where
+        F: FnMut((K, V), CacheChange),
+        K: Borrow<Q>,
+        Q: Hash + Eq,
     {
-        self.values.retain(|value| {
-            let retain = to_retain.contains(&value);
-            if !retain {
-                on_change(*value, CacheChange::Removal)
+        self.values.retain(|key, value| {
+            match to_update.get(key.borrow()) {
+                // New value matches old value. Do nothing.
+                Some(new_value) if new_value == value => true,
+                // If the new value isn't equal to the old value, overwrite
+                // the old value.
+                Some(new_value) =>  {
+                    let old_value = mem::replace(value, new_value.clone());
+                    on_change((*key, old_value), CacheChange::Modification);
+                    true
+                },
+                // Key doesn't exist, remove it from the map.
+                None => {
+                    on_change((*key, value.clone()), CacheChange::Removal);
+                    false
+                }
             }
-            retain
         });
         self.reset_on_next_modification = false;
     }
 }
 
+impl<'a, K, V> IntoIterator for &'a Cache<K, V>
```

Comment:
Is this actually used? If so, we probably don't need the `as_map()`.  if not, please remove.

Reply:
This is now used instead of `as_map` in the one place we actually iterate over the items of a cache. Will remove `as_map`.

Reply:
In the near future we should find some way  to auto-derive these for wrappers.